internal/users/tempentries/groups.go

```diff
@@ -0,0 +1,179 @@
+		}
+
+		// To avoid races where a group with this GID is created by some NSS source after we checked, we register this
+		// GID in our NSS handler and then check if another group with the same GID exists in the system. This way we
+		// can guarantee that the GID is unique, under the assumption that other NSS sources don't add groups with a GID
+		// that we already registered (if they do, there's nothing we can do about it).
+		var tmpID string
+		tmpID, cleanup, err = r.addTemporaryGroup(gid, name)
+		if err != nil {
+			return 0, nil, fmt.Errorf("could not register temporary group: %w", err)
+		}
+
+		unique, err := r.uniqueNameAndGID(name, gid, tmpID)
+		if err != nil {
+			if cleanupErr := cleanup(); cleanupErr != nil {
+				err = errors.Join(err, cleanupErr)
+			}
+			return 0, nil, fmt.Errorf("could not check if GID %d is unique: %w", gid, err)
+		}
+		if unique {
+			break
+		}
+
+		// If the GID is not unique, remove the temporary group and generate a new one in the next iteration.
+		if err := cleanup(); err != nil {
+			return 0, nil, fmt.Errorf("could not remove temporary group %q: %w", name, err)
+		}
+	}
+
+	log.Debugf(context.Background(), "Registered group %q with GID %d", name, gid)
+	return gid, cleanup, nil
+}
+
+func (r *temporaryGroupRecords) uniqueNameAndGID(name string, gid uint32, tmpID string) (bool, error) {
+	entries, err := localentries.GetGroupEntries()
+	if err != nil {
+		return false, err
+	}
+	for _, entry := range entries {
+		if entry.Name == name && entry.Passwd != tmpID {
+			// A group with the same name already exists, we can't register this temporary group.
+			log.Debugf(context.Background(), "Name %q already in use by GID %d", name, entry.GID)
+			return false, fmt.Errorf("group %q already exists", name)
+		}
+
+		if entry.GID == gid && entry.Passwd != tmpID {
+			log.Debugf(context.Background(), "GID %d already in use by group %q, generating a new one", gid, entry.Name)
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *temporaryGroupRecords) addTemporaryGroup(gid uint32, name string) (tmpID string, cleanup func() error, err error) {
+	r.rwMutex.Lock()
+	defer r.rwMutex.Unlock()
+
+	// Generate a 64 character (32 bytes in hex) random ID which we store in the passwd field of the temporary group
+	// record to be able to identify it in isUniqueGID.
+	bytes := make([]byte, 32)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", nil, fmt.Errorf("failed to generate random name: %w", err)
+	}
+	tmpID = fmt.Sprintf("authd-temp-group-%x", bytes)
+
+	r.groups[gid] = groupRecord{name: name, gid: gid, passwd: tmpID}
+	r.gidByName[name] = gid
+
+	cleanup = func() error { return r.deleteTemporaryGroup(gid) }
+
+	return tmpID, cleanup, nil
+}
+
+func (r *temporaryGroupRecords) deleteTemporaryGroup(gid uint32) error {
+	r.rwMutex.Lock()
+	defer r.rwMutex.Unlock()
+
+	group, ok := r.groups[gid]
+	if !ok {
+		return fmt.Errorf("temporary group with GID %d does not exist", gid)
+	}
```

Comment:
While having the cleanup function returning an error isn't a problem per se, I don't feel it's necessary either here... IMHO we could just ignore it, no?

Reply:
That's just for robustness: If we ever see that error, it would mean that there's a bug in our code which results in temporary groups not being added or removed the way they should.

Reply:
Sure I see that, but maybe we can just log in the function and that's it?

As it would make the calling code cleaner I think.

Reply:
> Sure I see that, but maybe we can just log in the function and that's it?

That makes it a lot less discoverable but ok, I'll change it

Reply:
done